plugins/uefi-dbx/fu-self-test.c

```diff
@@ -45,17 +54,658 @@ fu_efi_image_func(void)
+{
+	/* run though an update */
+
+	FuTestCase *tc = (FuTestCase *)user_data;
+	gboolean ret;
+	FuContext *ctx = fixture->ctx;
+	GPtrArray *devs = NULL;
+	FuDevice *dev = NULL;
+	FuEfivars *efivars = fu_context_get_efivars(ctx);
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GInputStream) stream_fw = NULL;
+	g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+	g_autoptr(FuPlugin) plugin = fu_plugin_new_from_gtype(fu_uefi_dbx_plugin_get_type(), ctx);
+
+	if (tc->with_snapd && !fixture->mock_snapd_available) {
+		g_test_skip("mock snapd not available");
+		return;
+	}
+
+	if (!fu_test_mock_dbx_efivars(efivars, &error)) {
+		if (g_error_matches(error, G_FILE_ERROR, G_FILE_ERROR_NOENT)) {
+			g_test_skip("test assets unavailable");
+			return;
+		}
+		g_assert_no_error(error);
+	}
+
+	ret = fu_plugin_runner_startup(plugin, progress, &error);
+	g_assert_no_error(error);
+	g_assert_true(ret);
+
+	ret = fu_plugin_runner_coldplug(plugin, progress, &error);
+	g_assert_no_error(error);
+	g_assert_true(ret);
+
+	devs = fu_plugin_get_devices(plugin);
+	g_assert_nonnull(devs);
+	g_assert_cmpuint(devs->len, ==, 1);
+	dev = g_ptr_array_index(devs, 0);
+
+	stream_fw = fu_test_mock_dbx_update_stream();
+	ret = fu_plugin_runner_write_firmware(plugin,
+					      dev,
+					      stream_fw,
+					      progress,
+					      /* skip verification of ESP binaries*/
+					      FWUPD_INSTALL_FLAG_FORCE,
+					      &error);
+	g_assert_no_error(error);
+	g_assert_true(ret);
+	g_assert_true(fu_device_has_flag(dev, FWUPD_DEVICE_FLAG_NEEDS_REBOOT));
+
+	/* this is normally invoked by FuEngine */
+	ret = fu_device_cleanup(dev, progress, 0, &error);
+	g_assert_true(ret);
+	g_assert_no_error(error);
+
+	if (tc->with_snapd) {
+		fu_self_test_mock_snapd_assert_calls(fixture,
+						     (FuTestSnapdCalls){
+							 .startup = 1,
+							 .prepare = 1,
+							 .cleanup = 1,
+						     });
+	}
+}
+
+static void
+fu_uefi_dbx_test_plugin_failed_update(FuTestFixture *fixture, gconstpointer user_data)
+{
+	/* update which fails at either write or cleanup steps, this test can only
+	 * properly mock the environment when using snapd integration */
+
+	FuTestCase *tc = (FuTestCase *)user_data;
+	gboolean ret;
+	FuContext *ctx = fixture->ctx;
+	GPtrArray *devs = NULL;
+	FuDevice *dev = NULL;
+	FuEfivars *efivars = fu_context_get_efivars(ctx);
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GInputStream) stream_fw = NULL;
+	g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+	g_autoptr(FuPlugin) plugin = fu_plugin_new_from_gtype(fu_uefi_dbx_plugin_get_type(), ctx);
+
+	if (!tc->with_snapd) {
+		g_test_skip("only supports snapd integration variant");
+		return;
+	}
+
+	if (!fixture->mock_snapd_available) {
+		g_test_skip("mock snapd not available");
+		return;
+	}
+
+	if (!fu_test_mock_dbx_efivars(efivars, &error)) {
+		if (g_error_matches(error, G_FILE_ERROR, G_FILE_ERROR_NOENT)) {
+			g_test_skip("test assets unavailable");
+			return;
+		}
+		g_assert_no_error(error);
```

Comment:
I think you want this assertion outside of the failure case (you want to make sure that no errors were asserted if `fu_test_mock_dbx_efivars` returned `TRUE`.